runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -119,20 +124,32 @@ public Object getPayload(CorfuRuntime runtime) {
     @Override
     public synchronized void releaseBuffer() {
         if (serializedCache != null) {
-            serializedCache.release();
-            if (serializedCache.refCnt() == 0) {
+            serializedCache.buffer.release();
+            if (serializedCache.buffer.refCnt() == 0) {
                 serializedCache = null;
             }
         }
     }
 
     @Override
-    public synchronized void acquireBuffer() {
-        if (serializedCache == null) {
-            serializedCache = Unpooled.buffer();
-            doSerializeInternal(serializedCache);
+    public synchronized void acquireBuffer(boolean metadata) {
+        if (serializedCache != null) {
+            if (metadata) {
+                serializedCache.buffer.resetReaderIndex();
+                serializedCache.buffer.writerIndex(serializedCache.metadataOffset);
+                doSerializeMetadataInternal(serializedCache.buffer);
+            }
+            serializedCache.buffer.retain();
+            return;
+        }
+
+        ByteBuf buf = Unpooled.buffer();
+        if (metadata) {
+            int metadataOffset = doSerializeInternal(buf);
+            serializedCache = new SerializedCache(buf, metadataOffset);
         } else {
-            serializedCache.retain();
+            doSerializePayloadInternal(buf);
+            serializedCache = new SerializedCache(buf, buf.writerIndex());
```

Comment:
This is a perfect place to use a CompositeByteBuf. The logic of maintaining an offset and overwriting can be greatly simplified. You can also remove the `SerializedCache` class

```java

        CompositeByteBuf serializationCache = Unpooled.compositeBuffer();
        ByteBuf metadata = Unpooled.buffer();
        ByteBuf payload = Unpooled.buffer();
        serializationCache.addComponents(metadata, payload);
```

Reply:
@Maithem Using CompositeByteBuf is exactly same idea that I was initially thinking, but later I changed to simply use one a buffer and offset, because the major purpose of CompositeByteBuf is to provide a single reader/writer index for multiple buffers, but in our case, we mostly want to overwrite the metadata part, which gives two options: 1) using the single reader/writer index that CompositeByteBuf provide, which is same as the current approach, and it's cleaner to just use the normal buffer; 2) maniputate the internal metadata sub-component. This would make confusion on the indexes, and also it's not the standard pattern of using CompositeByteBuf. So I changed to simply use a normal buffer and offset.

Reply:
I thought that was one of the main use cases of the CompositeByteBuf: allowing multiple buffers to be manipulated independently with the ability to append them to each other cheaply.

https://livebook.manning.com/book/netty-in-action/chapter-5/47

Reply:
So I assume you're referring to the second approahc that manipulates internal metadata sub-component without the offset variable? I that case, if we changed the metadata component, then you also need to change the unified writer index to payload componet writer index + new metadata size. I think this kind of manipulating writer index after write finishes seems error prone and would cause confusion, and it's not clean as having a normal buffer...